mitodl/ocw-studio
# Subscripts and superscripts

#### Pre-Flight checklist


#### What are the relevant tickets?
#1061 

#### What's this PR do?
This PR adds support for subscripts and superscripts in markdown via inclusion of raw `<sub>` and `<sup>` tags. Arbitrary HTML is not permitted: only HTML tags whitelisted in the markdown widget config's `allow_html` list will be permitted. Additionally, the only non-markdown tags that our CKEditor setup will generate are sub/sup, and only if the tags are allowed.

This approach turned out to not be as quite "works out of the box" as I originally thought. Two issues came up, both having to do with the fact that while MD --> HTML conversion is well specified by CommonMark, there is no specification for  HTML --> MD conversion (which we need when extracting data from CKEditor). The two issues were:

1. **`<sup>` at start of paragraph:** Markdown literally cannot handle this. HTML at the beginning of a line creates an HTML block, and the remainder of the line is parsed as raw HTML rather than Markdown. So the two lines

    ```md
    <sup>1</sup>:  This renders raw html with **asterisks**

    x<sup>1</sup> This renders as markdown with **bold**.
    ```

    end up behaving very differently.
2. **MD inside html tags**: The default behavior for Turndown (our html->md converter) when including raw HTML in markdown is to not process the content of the HTML *at all*. But HTML tags are generally allowed to contain markdown[^1] content, and it's important to convert that content to markdown.

[^1]: This is absolutely true for [markdown inline html](https://spec.commonmark.org/0.30/#raw-html). The story is a little more complicated for [markdown block html](https://spec.commonmark.org/0.30/#html-blocks).

#### How should this be manually tested?
1. Load the course-v2 config from https://github.com/mitodl/ocw-hugo-projects/pull/234 into your local studio
2. Create a page and add subscripts/superscripts. Some things that are worth trying are subscripts/superscripts
    - at the beginning of a line
    - in links and resource links
    - in tables
    - bold inside the subscript/superscript *though we do not encourage authors to do this, ckeditor will not prevent it.*
3. Check the markdown in admin panel. It should look correct.
4. Check that the markdown renders correctly in ocw-hugo-themes using branch from https://github.com/mitodl/ocw-hugo-themes/pull/1004

#### Screenshots

<img width="1279" alt="Screen Shot 2022-11-30 at 9 05 59 AM" src="https://user-images.githubusercontent.com/9010790/204816853-1a0a69de-8daa-47a4-adbc-4637f2f6e7d5.png">

#### Other Context

HQ discussion: https://github.com/mitodl/hq/discussions/303
